File: mobile/src/features/feed/services/suggestionService.ts
// src/features/feed/services/suggestionService.ts
// User Suggestion API Service
// Backend: GET /api/users/suggested

import { apiClient } from '@core/api/client';
import { API_ENDPOINTS } from '@core/api/endpoints';

/**
 * Suggested User Response from Backend
 * Backend: SuggestedUserResponse.java
 *
 * Replaces:
 * - mockExperts.ts (MOCK_SUGGESTED_EXPERTS)
 * - NoFollowingEmptyState.types.ts (MOCK_SUGGESTED_EXPERTS)
 */
export interface SuggestedUserResponse {
  id: number;
  fullName: string;
  profession: string | null;
  avatarUrl: string | null;
  isVerified: boolean;
  isFollowing: boolean;
  followerCount: number;
}

/**
 * Get suggested users for current user
 *
 * @param limit - Number of suggestions (default: 8, max: 20)
 * @returns List of suggested users
 *
 * Backend:
 * - Endpoint: GET /api/users/suggested?limit={limit}
 * - Service: SuggestionService.java
 * - Algorithm: Profession (50%) + Engagement (30%) + Verified (20%)
 * - Cache: 5 minutes per user
 * - Rate Limit: 60 requests/hour
 *
 * Replaces: mockExperts.ts (MOCK_SUGGESTED_EXPERTS)
 */
export async function getSuggestedUsers(limit: number = 8): Promise<SuggestedUserResponse[]> {
  const response = await apiClient.get<SuggestedUserResponse[]>(API_ENDPOINTS.USER_SUGGESTIONS, {
    params: { limit },
  });
  return response.data;
}

File: backend/src/main/java/com/dengin/social/application/service/SuggestionService.java
package com.dengin.social.application.service;

import com.dengin.identity.domain.model.User;
import com.dengin.social.application.dto.SuggestedUserResponse;
import com.dengin.social.domain.repository.FollowRepository;
import com.dengin.social.domain.service.UserSuggestionService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

/**
 * User Suggestion Application Service
 * 
 * Orchestrates user suggestion use cases:
 * - Get suggested users for current user
 * - Map domain models to DTOs
 * - Cache management
 * 
 * Layer: Application Service (coordinates domain services)
 * Caching: 5 minutes per user (suggestions may change frequently)
 * Transaction: Read-only
 * 
 * Replaces Mobile Mocks:
 * - mockExperts.ts (MOCK_SUGGESTED_EXPERTS)
 * - NoFollowingEmptyState.types.ts (MOCK_SUGGESTED_EXPERTS)
 * 
 * @see UserSuggestionService
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class SuggestionService {
    
    private final UserSuggestionService userSuggestionService;
    private final FollowRepository followRepository;
    
    private static final int DEFAULT_LIMIT = 8;
    
    /**
     * Get suggested users for current user
     * 
     * @param currentUserId Current user ID
     * @param limit         Number of suggestions (default: 8)
     * @return List of suggested user responses
     */
    @Transactional(readOnly = true)
    @Cacheable(
        value = "user-suggestions",
        key = "#currentUserId + ':' + #limit",
        unless = "#result == null || #result.isEmpty()"
    )
    public List<SuggestedUserResponse> getSuggestedUsers(Long currentUserId, int limit) {
        log.debug("Getting {} suggested users for user {}", limit, currentUserId);
        
        List<UserSuggestionService.ScoredUser> scoredUsers = 
            userSuggestionService.getSuggestedUsers(currentUserId, limit);
        
        return scoredUsers.stream()
            .map(this::mapToResponse)
            .collect(Collectors.toList());
    }
    
    /**
     * Get suggested users with default limit (8)
     */
    public List<SuggestedUserResponse> getSuggestedUsers(Long currentUserId) {
        return getSuggestedUsers(currentUserId, DEFAULT_LIMIT);
    }
    
    /**
     * Map ScoredUser to SuggestedUserResponse DTO
     */
    private SuggestedUserResponse mapToResponse(UserSuggestionService.ScoredUser scoredUser) {
        User user = scoredUser.user;
        
        SuggestedUserResponse response = new SuggestedUserResponse();
        response.setId(user.getId());
        response.setFullName(user.getFullName());
        response.setProfession(user.getProfession() != null ? user.getProfession().getName() : null);
        response.setAvatarUrl(user.getAvatarUrl());
        response.setVerified(user.isVerified());
        response.setFollowing(false); // Always false for new suggestions
        
        // Get follower count
        long followerCount = followRepository.countByFollowingId(user.getId());
        response.setFollowerCount(followerCount);
        
        // Optional: Include relevance score for debugging
        // response.setRelevanceScore(scoredUser.score);
        
        return response;
    }
}
